### PR TITLE
(fix) DBT metadata ingest can add metadata to existing dataplatforms

### PR DESCRIPTION
**Pull request by luck02**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

Driver: https://datahubspace.slack.com/archives/CUMUWQU66/p1619190330199800

TL/DR: DBT metadata will likely be loaded to provide DBT metadata (tags and lineage) and we will want the dbt ingestion tooling to enhance existing data.  That data will likely be loaded from another platform.

IE: 
```
urn:li:dataset:(urn:li:dataPlatform:dbt,pagila.public.payment_p2020_01,PROD)
vs.
urn:li:dataset:(urn:li:dataPlatform:postgresql,pagila.public.payment_p2020_01,PROD)
```

We will provide either the ability to specify a target_platform or infer it from the dbt metadata.

### CODE REVIEW

File: metadata-ingestion/README.md [hunk 1]
```diff
@@ -483,13 +483,18 @@ Pull metadata from DBT output files:
 - [dbt catalog file](https://docs.getdbt.com/reference/artifacts/catalog-json)
   - This file contains schema data.
   - DBT does not record schema data for Ephemeral models, as such datahub will show Ephemeral models in the lineage, however there will be no associated schema for Ephemeral models
+  - Not required if platform configuration provided.
+- platform: Optional.
+  - If provided the ingestion tool will not attempt to load a catalog file.
+  - In this case we assume DBT metadata is enriching an existing data platform.
```
Comment (jjoyce0510) [inline metadata-ingestion/README.md:489]: In which case?

**Comment by jjoyce0510** [comment]
Thanks for the contribution!

If I'm understanding correctly, sometimes we want to say that the DataPlatform associated with a Dataset is "dbt" as opposed to the warehouse that DBT is writing back to? (ie. Snowflake, Redshift)... I'm wondering in what cases we won't have the platform where the data is actually materialized... that is, when using "dbt" as the platform would be appropriate. 

@luck02 Do you mind elaborating on motivation for the PR inside the description? 

Is there any way we can parse this information from the DBT configuration files such that even if the "platform" isn't configured we can still infer it?

**Comment by luck02** [comment]
> Thanks for the contribution!
> 
> If I'm understanding correctly, sometimes we want to say that the DataPlatform associated with a Dataset is "dbt" as opposed to the warehouse that DBT is writing back to? (ie. Snowflake, Redshift)... I'm wondering in what cases we won't have the platform where the data is actually materialized... that is, when using "dbt" as the platform would be appropriate.
> 
> @luck02 Do you mind elaborating on motivation for the PR inside the description?
> 
> Is there any way we can parse this information from the DBT configuration files such that even if the "platform" isn't configured we can still infer it?

hi @jjoyce0510  the motivation is here: https://github.com/linkedin/datahub/pull/2483#issuecomment-830914549 essentially most organizations would want the lineage and tags from the DBT metadata but still load the raw assets (schema etc) from the data store itself.

File: metadata-ingestion/README.md (hunk 1 above)
Comment (luck02) [reply metadata-ingestion/README.md]: improved the wording, marking as resolve.

**Comment by jjoyce0510** [comment]
Thanks for the detailed explanation Gary, much appreciated. 

Going to focus on answering the questions around 'platform' since I think it's the most relevant part to this PR.

> A question for the 'core' datahub team... Does this team see platform as a stable concept? IE is it expected to be internally defined and implicit indefinitely or is it more of an unknown at this time? One additional question, is there a comprehensive definition of what platform means in Datahub? Maybe I'm misinterpreting the term?

As we see it, Data Platforms can represent arbitrary third-party systems: data storage, compute, streaming, visualization, etc. The key thing about platforms is that they denote the _technology_ itself, as opposed to a particular cluster, schema, or table (logical partition) beyond that. 

With respect to the redshift example you provided, I image that "platform" would remain the same for datasets across the 2 redshift clusters (it would be _redshift_). I think we'd want to add an additional qualification or "label" capable of capturing this cluster information. One possibility to do so would be to change the existing FabricType in the Urns from enum to string type, allowing arbitrary cluster / env / label information to be encoded. From what I've seen, some people actually encode the cluster information in the dataset name itself (though this can become very ugly in practice). 

I think it's important to capture the cluster-level metadata as you pointed out. I'll loop back around with the team on this topic & keep you updated on the thinking!

**Comment by luck02** [comment]
@jjoyce0510  thanks for the feedback! 

> I think it's important to capture the cluster-level metadata as you pointed out. I'll loop back around with the team on this topic & keep you updated on the thinking!

I think for the purposes of this effort I will proceed with providing platform config as I don't think I can really guarantee inferring it will work in all cases.  I view this as an incremental improvement and we can always change it later if it becomes problematic. 

If we have a platform defined then we won't load a catalog as that would likely overwrite existing data but we'll still get DBT metadata.  

Presumably if the dbt catalog data is non-usefull we can drop it in the future.

Thanks again!

**Comment by luck02** [comment]
Another concern comes up.  DBT keeps the materialization info for all sources (edge nodes) in the catalog so we probably want to require both for now.  

A later iteration could probably dispense with the catalog when DBT is being used to enhance an existing platform. IE we could infer at least some of the materialization type from the DB.   

I think this is future work in the name of keeping the PR focused.

**Comment by luck02** [comment]
@jjoyce0510 / @hsheth2 

This is final state right now.  Once this is merged in I'll set up another PR with the intention of adding:

* Minimal unit tests as we now have a branch in the code which significantly changes how the ingestion works.  I don't want a separate integration tests for the two scenarios.
* Add a DBT test to the source DBT so we can gracefull handle tests re: https://datahubspace.slack.com/archives/CUMUWQU66/p1620130515310200 
* Add a DBT incremental model to ensure we capture that effectively.

I'm still uncertain if we should try and determine data platform from the metadata.  I'd prefer to address that separately, at least this way we can get more people pulling their DBT data into datahub in the near future :) 

In any case if this PR get a thumbsup let me know if I should go ahead and squash my commits prior to merge :) 

Here's a screenshot of a lineage page:
image

Here's a screenshot of the schema we get from the postgres adapter:

image

And here's a screenshot with DBT properties:
image

File: metadata-ingestion/README.md [hunk 2]
```diff
@@ -483,13 +483,16 @@ Pull metadata from DBT output files:
...
+- target_platform: Optional
+  - In this case we use DBT metadata to enrich an existing data platform with DBT sourced lineage and metadata.
 
 ```yml
 source:
   type: "dbt"
   config:
     manifest_path: "./path/dbt/manifest_file.json"
     catalog_path: "./path/dbt/catalog_file.json"
+    target_platform: "dbt" # optional eg postgresql, snowflake etc.
```
Comment (hsheth2) [inline metadata-ingestion/README.md:495]: It should be `postgres` instead of `postgresql`. Also, can we add a link to https://github.com/linkedin/datahub/blob/master/gms/impl/src/main/resources/DataPlatformInfo.json.

File: metadata-ingestion/src/datahub/ingestion/source/dbt.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ class DBTConfig(ConfigModel):
     manifest_path: str
     catalog_path: str
     env: str = "PROD"
+    target_platform: str = "dbt"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/dbt.py:39]: I'm a bit unsure on this config being optional - when does it make sense to use "dbt" as the platform?

Maybe we can have `target_platform` required and then also add a `load_schemas` flag to the config?

**Review by hsheth2 — COMMENTED** [review]
> I think for the purposes of this effort I will proceed with providing platform config as I don't think I can really guarantee inferring it will work in all cases. I view this as an incremental improvement and we can always change it later if we determine there's a better route. At least we'll get this working for most people.

Sounds good

Re materialization information - I think it makes sense to keep it simple by continuing to require the catalog file as well. I'm assuming that the catalog and manifest files are often generated side-by-side, so it won't be a large inconvenience. 

Testing will be important for this, but I agree that we can push that to a follow-up PR.

**Comment by remisalmon** [comment]
@luck02 thank you for your work on this PR! I have tested it and `target_platform` is working as expected, with only one exception:

By default dbt will name its source tables after the table names form the db, but it is possible to set a different source table name using the `identifier` parameter (in that case `identifier` is the table name in the db and `name` the source table name within dbt, see https://docs.getdbt.com/reference/resource-properties/identifier for a better explanation).
In that case the dbt ingestion creates new unwanted tables `db.name` in DataHub, instead of linking to the correct `db.identifier`.

This can be fixed by replacing `name = node['name']` by `name = node['identifier'] if 'identifier' in node else node['name']` in both `extract_dbt_entities()` and `get_upstreams()`.

Let me know if that makes sense, I can open another PR if needed.

**Comment by luck02** [comment]
> This can be fixed by replacing `name = node['name']` by `name = node['identifier'] if 'identifier' in node else node['name']` in both `extract_dbt_entities()` and `get_upstreams()`.

Hi @remisalmon 

I'm leaning towards addressing this in a different PR, my concern is I'd want to add in your situation to the test DBT project before adding in additional complexity (https://github.com/luck02/sample-dbt) and additionally I'm really time constrained at the moment :'( 

It would help me understand what you're doing if you were to modify that project with an identifier :)  

That being said you're also welcome to open a PR, if not I'll try to roll it into the next PR I open against this component.

Thank you for bringing this up!

**Comment by luck02** [comment]
force pushed after rebase.

**Comment by luck02** [comment]
Squashed commits and followed commit guidelines.

@hsheth2 can you take a peek?  Thanks!

**Review by hsheth2 — APPROVED** [review]
LGTM - thanks @luck02!

**Review by shirshanka — APPROVED** [review]
LGTM! Thanks @hsheth2 for the review and @luck02 for powering through on the weekends :)
